icu4c/source/test/intltest/canittst.cpp

```diff
@@ -332,4 +333,15 @@ void CanonicalIteratorTest::TestAPI() {
   }
 }
 
+void CanonicalIteratorTest::TestICU22642() {
```

Comment:
Without the bugfix this test case will just continue running until it eventually exhausts all resources of the machine it's running on. That's not useful as a unit test.

Reply:
but we want to ensure there are no regression if someone change the code later and reintroudce the problem.

Reply:
Yes, it's valuable to have a unit test, but the test you have written here now isn't a useful unit test. A useful unit test is something that fails with an error message when some unexpected change causes the thing being tested to no longer behave as expected, not something that continues running until all resources are exhausted and then crashes.

If you can't think of a way of writing a useful unit test for this, we'll simply have to rely on the fuzz test that found the problem this time to find the problem again if it reappears in the future.

Reply:
The broken behavior is it hang, how other way we can test it won't hang?

Reply:
I see both of your points.  I worry that if somebody checks in a change that causes this problem to happen again, what'll happen is that none of the CI checks will finish.  It'll be a while before we know they're stuck, and we won't know what to do for a while and will probably think something's wrong with the CI stuff.

What you might be able to do in this test is spawn two threads-- one that triggers the problem and another one that waits some prescribed amount of time for the first one to finish and that kills it and reports an error if it doesn't.

Reply:
> What you might be able to do in this test is spawn two threads-- one that triggers the problem and another one that waits some prescribed amount of time for the first one to finish and that kills it and reports an error if it doesn't.

I don't think this is a very good idea, for two reasons:

1. Introducing threading increases the complexity of the testing code significantly, introducing new potential sources of failures unrelated to the code actually being tested and also making it harder to run this in a debugger to see what's going on when something fails.
2. The speed of execution varies wildly between different execution environments and it'd be nigh impossible to pick a reasonable amount of time that isn't either far too long when run locally on a fast workstation or far too short when run on our slow CI pipelines.

Reply:
>and that kills it and reports an error if it doesn't.
My code currently does not "kills it"

@richgillam  Would it be ok for me to listen to @roubert by just removing the test and let the fuzzer do it's job to verify?

Reply:
Yeah, I'm fine with that.

Reply:
ok, I remove the test per @roubert . PTAL